### ISSUE TITLE
SagePay Payment missing a description

### ISSUE THREAD

**Issue by catharsisjelly** [silverstripe/silverstripe-omnipay#153]
Hi there,

I am trying to implement this for SagePay and the request comes back (from SagePay) telling em that it requires a description ([error code 3013](http://www.sagepay.co.uk/support/error-codes/malformed-description-missing)). The data that comes to the module is sent via the Order form sent to it via the `OrderActionsForm` of `silvershop/core`. So is the solution to add in a faked description into the order somehow?

Not sure where this issue should go so forgive me if you don't think it's a problem with the module

**Comment by dhensby** [comment]
Have you raised this with https://github.com/silvershop/silvershop-core because they are interfacing with this module so I'd assume it's for them to set the description

**Comment by bummzack** [comment]
@dhensby I think that's debatable. SilverShop doesn't really know each required field of every payment-gateway. This is also going to be an issue for anybody that uses this module directly and just supplies the minimal amount of data.

This can be easily solved with a DataExtension though. I'll post an example in a sec

**Comment by bummzack** [comment]
Hmm, I stand corrected. I thought it would be as simple as implementing `onBeforePurchase` and `onBeforeAuthorize` hooks on an extension and add the `description` to the gateway-data.

That's still a valid approach, but there's currently no easy and reliable way to get the current order within that hook.

I think it would be fine to add a public accessor to get the Payment from PaymentServices. So that you can get the current associated payment within the hook and with that you'll also have access to `Order` when in the shop context.

Another possible approach would be to modifiy checkout fields for shop (via `updateConfirmationForm` hook) and add a "description" field there

**Comment by bummzack** [comment]
I've given this some thought and I don't think adding the fields to the form in SilverShop is a good approach. Form values can easily be tinkered with and if you're dealing with a redirect gateway, it goes against the philosophy of omnipay to supply different values for `purchase` and `completePurchase`.

I think the best approach is to use an extension on `PaymentService` that implements `onBeforePurchase` and `onBeforeAuthorize`. 

@catharsisjelly 
Create an extension, like this:
```php
<?php
class SagePayExtension extends Extension
{
    public function onBeforePurchase(&$gatewayData)
    {
        $gatewayData['description'] = 'test';
    }

    public function onBeforeAuthorize(&$gatewayData)
    {
        $gatewayData['description'] = 'test';
    }
}
```
Add the extension to `PaymentService` by adding something like this to your config:

```yml
SilverStripe\Omnipay\Service\PaymentService:
  extensions:
    - SagePayExtension
```
Please check, if that fixes your issue.

Of course `test` isn't a good description. And as I've written before, there is currently not a good way to access Order information in this hook. I'll add an accessor to the Payment instance on `PaymentService`, so that the payment can be accessed within these hooks.

**Comment by bummzack** [comment]
I've added the `getPayment` method to `PaymentService` (currently in the master branch). With that in place, here's a more realistic example of an extension:

```php
<?php
class SagePayExtension extends Extension
{
    public function onBeforePurchase(&$gatewayData)
    {
        $this->addDescription($gatewayData);
    }

    public function onBeforeAuthorize(&$gatewayData)
    {
        $this->addDescription($gatewayData);
    }

    private function addDescription(&$gatewayData)
    {
        $payment = $this->owner->getPayment();
        if ($payment->Gateway == 'SagePay_Direct' || $payment->Gateway == 'SagePay_Server') {
            if ($order = $payment->Order()) {
                $gatewayData['description'] = $order->Reference;
            } else {
                $gatewayData['description'] = $payment->Identifier;
            }
        }
    }
}
```

**Comment by catharsisjelly** [comment]
@bummzack Thanks for the response, this extension works as you expect, thanks for making the change. Is there anything I can contribute towards this?

**Comment by bummzack** [comment]
@catharsisjelly I'm not sure if I should release this extension as part of shop… I imagine that the contents of `description` could vary easily depending on client demand…

I'm not sure if we had anybody using sage-pay with shop yet. So testing and confirming that it works is enough of a contribution for me :)

**Comment by catharsisjelly** [comment]
@bummzack Great, be aware that the description for SagePay must be < 100 characters, if I find any more issues while I am implementing this I'll log more issues. In theory the description shows up as part of the transaction in the SagePay portal so might be useful to be able to manipulate what goes in there but default to an order number or something to that effect

**Comment by catharsisjelly** [comment]
PR Reference #163

**Comment by bummzack** [comment]
Closed via #163 and released in 2.2.0
